### ISSUE TITLE
warn transmute about non-exhaustives

### ISSUE THREAD

**Issue by KisaragiEffective** [rust-lang/rust-clippy#10290]
### What it does

This lint emit warnings in following condition.

let `T` is before-transmute type, and `U` is after-transmute type.
if valid range of `T` does not covered by `U`, this is UB.
This lint bounds `T` to integral types or `enum` with `#[repr(u_)]` or `#[repr(i_)]` and its all variants does not have fields, and `U` to `#[repr(u_)]` or `#[repr(i_)]`. 

TODO: If `T` is `enum` with `#[repr(C)]` and its all variants does not have fields, treat it as `#[repr(u<GuessedWidth>)]`.

More spefically:

|           T             |        U         |
|:-----------------------:|:----------------:|
|`u8`<br>`#[repr(u8)]`   　|`#[repr(u8)]`     |
|`u8`<br>`#[repr(u8)]`   　|`#[repr(i8)]`     |
|`i8`<br>`#[repr(i8)]`   　|`#[repr(u8)]`     |
|`i8`<br>`#[repr(i8)]`   　|`#[repr(i8)]`     |
|`u16`<br>`#[repr(u16)]`   |`#[repr(u16)]`     |
|`u16`<br>`#[repr(u16)]`   |`#[repr(i16)]`     |
|`i16`<br>`#[repr(i16)]`   |`#[repr(u16)]`     |
|`i16`<br>`#[repr(i16)]`   |`#[repr(i16)]`     |
|`u32`<br>`#[repr(u32)]`   |`#[repr(u32)]`     |
|`u32`<br>`#[repr(u32)]`   |`#[repr(i32)]`     |
|`i32`<br>`#[repr(i32)]`   |`#[repr(u32)]`     |
|`i32`<br>`#[repr(i32)]`   |`#[repr(i32)]`     |
|`u64`<br>`#[repr(u64)]`   |`#[repr(u64)]`     |
|`u64`<br>`#[repr(u64)]`   |`#[repr(i64)]`     |
|`i64`<br>`#[repr(i64)]`   |`#[repr(u64)]`     |
|`i64`<br>`#[repr(i64)]`   |`#[repr(i64)]`     |

If their value may not overwrap (i.e. integer to one-element enum), emits warning.
If their value never overwrap (i.e. `Foo { One = 1 }` and `Bar { Two = 2 }`), transmuting between them is 100% UB and emits error (this should be given another lint name, like `transmute_non_overwrapping_enums` under `correctness`)
If `U` is 100% exhaustive over integral possible values (i.e. `#[repr(u8)]` and having 256 elements), this lint does not fire.

### Lint Name

transmute_integer_to_out_of_enum

### Category

suspicious

### Advantage

- Avoids some form of UB.

### Example

```rust
#[repr(u8)]
enum Foo {
    Zero,
    One,
    Two,
}

let invalid = 3;
let ub: Foo = unsafe { std::mem::transmute(invalid) };
//                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Could be written as:

```rust
impl Foo {
    fn new(inner: u8) -> Option<Foo> {
        // match clause
    }
}
let invalid = 3;
let not_ub: Foo = Foo::new(invalid).expect_err("should not succeed")
```
